java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/commands/ProjectMetadataCommand.java

```diff
@@ -0,0 +1,180 @@
+                            FileObject f = (FileObject)t;
+                            writer.value(f == null ? null : f.getPath());
+                        }
+
+                        @Override
+                        public T read(JsonReader reader) throws IOException {
+                            if (reader.peek() == JsonToken.NULL) {
+                                reader.nextNull();
+                                return null;
+                            } else {
+                                String s = reader.nextString();
+                                if (s == null) {
+                                    return null;
+                                }
+                                FileObject fo = FileUtil.toFileObject(Paths.get(s).toFile());
+                                return (T)fo;
+                            }
+                        }
+                    };
+                }
+            })
+            .create();
+    }
+    
+    @Override
+    public List<CodeAction> getCodeActions(ResultIterator resultIterator, CodeActionParams params) throws Exception {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Set<String> getCommands() {
+        return COMMANDS;
+    }
+
+    @Override
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        if (arguments.size() < 1) {
+            throw new IllegalArgumentException("Expected at least project URI/path");
+        }
+        FileObject f = Utils.extractFileObject(arguments.get(0), gson);
+        Project p = FileOwnerQuery.getOwner(f);
+        if (p == null || p.getProjectDirectory() != f) {
+            throw new IllegalArgumentException("Not a project " + f);
+        }
+        String artifactType = null;
+        ProjectActionContext ctx = null;
+        
+        if (arguments.size() > 1) {
+            // 2nd parameter is the project action
+            Object o = arguments.get(1);
+            if (!(o instanceof JsonPrimitive)) {
+                throw new IllegalArgumentException("String or null expected as parameter #3, got " + o);
+            }
+            ctx = ProjectActionContext.newBuilder(p).forProjectAction(((JsonPrimitive)o).getAsString()).context();
+        }
+        if (arguments.size() > 2) {
+            // 3rd parameter is the type of artifact
+            Object o = arguments.get(2);
+            if (!(o instanceof JsonPrimitive)) {
+                throw new IllegalArgumentException("String or null expected as parameter #2, got " + o);
+            }
+            artifactType = ((JsonPrimitive)o).getAsString();
+        }
+        ProjectArtifactsQuery.Filter filter = ProjectArtifactsQuery.newQuery(artifactType, null, ctx);
+        CompletableFuture result = new CompletableFuture();
+        METADATA_PROCESSOR.post(() -> {
+            try {
+                ArtifactsResult arts = ProjectArtifactsQuery.findArtifacts(p, filter);
```

Comment:
All the changes _just_ to get name of artifact to VSCode TypeScript side...

Reply:
erm. not just the name, but also the location of the potential output. Yes, it's large, but NetBeans Project APIs themselves lack a description of usual project metadata, some identifier etc. The location of to-be-compiled output could be useful for packaging steps in NB, too.